src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -262,15 +262,17 @@ static void adaptSegment(
 
     if (permanence < EPSILON)
     {
-      connections.destroySynapse(synapse);
+      connections.destroySynapse(synapses[i]);
+      // Synapses vector is modified in-place, so don't update `i`.
     }
     else
     {
-      connections.updateSynapsePermanence(synapse, permanence);
+      connections.updateSynapsePermanence(synapses[i], permanence);
```

Comment:
```
connections.updateSynapsePermanence(synapses[i++], permanence);
```

Although I think doing the inc on the next line is more explicit / easier to understand so probably leave as is.